### ISSUE TITLE
VM disk monitoring not working as expected

### ISSUE THREAD

**Issue by atodorov-storpool** [OpenNebula/one#4793]
**Description**
After enabling the workaround of #4791 something is working but it looks like the sent monitoring data from the remote hosts is not processed, despite the `Successfully monitored VM: $VM_ID` log message

**To Reproduce**
install opennebula and a remote KVM host
instantiate a VM in the default ssh SYSTEM datatore
check the VM disk size

**Expected behavior**
The VM disks size should be updated

**Details**
 - Affected Component: [Monitring]
 - Hypervisor: [KVM]
 - Version: [development]

**Additional context**
As far as I understand there are two ways to feed VM disk stats to OpenNebula:
1. on the frontend node - TM_MAD/monitor when called by for SYSTEM datastore which is not tagged for "at the host" monitoring (Cant twist my mind to name it "tagged for local monitoring"...)
2. On the hosts - TM_MAD/monitor_ds called by the remote monitoring probe

the monitoring message should be 
```
VM = [ ID = $VM_ID, MONITOR = "<base64>
DISK_SIZE=[ID=$DISK_ID,SIZE=$DISK_SIZE]
SNAPSHOT_SIZE = [ID=$SNAP_ID, DISK_ID=$DISK_ID, SIZE=$SNAP_SIZE]
</base64>" ]
```
The message is logged as successfully monitored by monitors:
```
Fri May 22 10:41:02 2020 [Z0][HMM][I]: Successfully monitored VM: 3
```
But in the VMs the disks sizes are 0:

```
[root@one512 ssh]# onevm show 3
VIRTUAL MACHINE 3 INFORMATION                                                   
ID                  : 3                   
NAME                : Ttylinux - KVM-3    
USER                : oneadmin            
GROUP               : oneadmin            
STATE               : ACTIVE              
LCM_STATE           : RUNNING             
LOCK                : None                
RESCHED             : No                  
HOST                : vs04                
CLUSTER ID          : 0                   
CLUSTER             : default             
START TIME          : 05/20 16:15:55      
END TIME            : -                   
DEPLOY ID           : 6d8f347e-fbce-4d26-9df5-d455d6a2c11b

VIRTUAL MACHINE MONITORING                                                      
CPU                 : 1.01                
MEMORY              : 106.1M              
NETTX               : 0K                  
NETRX               : 0K                  
DISKRDBYTES         : 13855320            
DISKRDIOPS          : 974                 
DISKWRBYTES         : 596992              
DISKWRIOPS          : 303                 
ID                  : 3                   
TIMESTAMP           : 1590133292          

PERMISSIONS                                                                     
OWNER               : um-                 
GROUP               : ---                 
OTHER               : ---                 

VM DISKS                                                                        
 ID DATASTORE  TARGET IMAGE                               SIZE      TYPE SAVE
  0 default    vda    Ttylinux - KVM                      -/200M    file   NO
  2 system     sda    1024M                               -/1024M   swap   NO
  1 -          hda    CONTEXT                             -/-       -       -
```
patched the monitor_ds script to save the base64 encoded string to a temp file
```
    echo -e "$vm_monitor" | base64 -w 0 - | tee /tmp/monitor_ssh${vm}.b64
```
And there is the following data provided:
```
[root@vs04 tmp]# cat /tmp/monitor_ssh3.b64 | base64 -d 

DISK_SIZE = [ ID=0, SIZE=27]
DISK_SIZE = [ ID=1, SIZE=1]
DISK_SIZE = [ ID=2, SIZE=1]
```

## Progress Status
- [ ] Branch created 
- [ ] Code committed to development branch
- [ ] Testing - QA
- [ ] Documentation 
- [ ] Release notes - resolved issues, compatibility, known issues
- [ ] Code committed to upstream release/hotfix branches
- [ ] Documentation committed to upstream release/hotfix branches

**Comment by atodorov-storpool** [comment]
for the monitoring on the frontend via TM_MAD/monitor there is no message logged at all

**Comment by atodorov-storpool** [comment]
managed at least to log the received message from remote monitoring but the VM's disk sizes are not updated yet

```
Fri May 22 11:25:28 2020 [Z0][MDP][D]: Received MONITOR_VM msg: VM = [ ID = 3, MONITOR = "CkRJU0tfU0laRSA9IFsgSUQ9MCwgU0laRT0yN10KRElTS19TSVpFID0gWyBJRD0xLCBTSVpFPTFdCkRJU0tfU0laRSA9IFsgSUQ9MiwgU0laRT0xXQo="]


VM = [ ID="3", DEPLOY_ID="6d8f347e-fbce-4d26-9df5-d455d6a2c11b", MONITOR="Q1BVPSIwLjAiCk1FTU9SWT0iMTA4NzQ4IgpORVRSWD0iMCIKTkVUVFg9IjAiCkRJU0tSREJZVEVTPSIxMzg1NTMyMCIKRElTS1dSQllURVM9IjYwMzEzNiIKRElTS1JESU9QUz0iOTc0IgpESVNLV1JJT1BTPSIzMDgiCg=="]
VM = [ ID="5", DEPLOY_ID="b4365ecb-a13d-4624-a138-fe2e877cae7d", MONITOR="Q1BVPSIxLjAxIgpNRU1PUlk9IjEwODU4NDAiCk5FVFJYPSIwIgpORVRUWD0iMCIKRElTS1JEQllURVM9IjEwNDEwOTgwOTIiCkRJU0tXUkJZVEVTPSIxMDMxNjgyMDQ4IgpESVNLUkRJT1BTPSIzNTA3MCIKRElTS1dSSU9QUz0iMjA0MzgiCg=="]
VM = [ ID="6", DEPLOY_ID="a1a6c694-e032-4940-a557-40f20ab5148e", MONITOR="Q1BVPSIxLjAxIgpNRU1PUlk9IjM4MDY3MiIKTkVUUlg9IjAiCk5FVFRYPSIwIgpESVNLUkRCWVRFUz0iMTYwNDIxMjI4IgpESVNLV1JCWVRFUz0iMjk0Nzk0MjQiCkRJU0tSRElPUFM9IjkwNzciCkRJU0tXUklPUFM9IjE2NjQiCg=="]
VM = [ ID="7", DEPLOY_ID="6eeceda6-6e25-4e2b-a447-a702fc770fb4", MONITOR="Q1BVPSIxLjAxIgpNRU1PUlk9IjM3NTE1NiIKTkVUUlg9IjAiCk5FVFRYPSIwIgpESVNLUkRCWVRFUz0iMTUzNzYzMTQwIgpESVNLV1JCWVRFUz0iMjczMDg1NDQiCkRJU0tSRElPUFM9Ijg3ODgiCkRJU0tXUklPUFM9IjE0NjYiCg=="]
Fri May 22 11:25:28 2020 [Z0][HMM][I]: Successfully monitored VM: 3
Fri May 22 11:25:28 2020 [Z0][HMM][I]: Successfully monitored VM: 3
Fri May 22 11:25:28 2020 [Z0][HMM][I]: Successfully monitored VM: 7
Fri May 22 11:25:28 2020 [Z0][HMM][I]: Successfully monitored VM: 6
Fri May 22 11:25:28 2020 [Z0][HMM][I]: Successfully monitored VM: 5
```

researching from where are the emty lines and are they causing issues...

**Comment by atodorov-storpool** [comment]
Hey @feldsam 

Do you have such issue with 5.12 beta? 

Please note that the format of the monitoring messages is changed in 5.12-beta: it is _MONITOR_ instead of _POLL_ and the _MONITOR_ content is base64 encoded

**Comment by atodorov-storpool** [comment]
hmm it is my confusion:
tm/monitor is with POLL
tm/monitor_ds is with MONITOR

Still neither working, though

**Comment by feldsam** [comment]
Hi @atodorov-storpool , I don't test 5.15 yet. What is `tm/monitor_ds` file? I don't have it in a 3par driver. Regarding monitoring, I have only issue with slow XPath when getting disks info from VM XML.

**Comment by atodorov-storpool** [comment]
Hi @feldsam ,

>  I don't test 5.15 yet. What is tm/monitor_ds file? I don't have it in a 3par driver.

This file is responsible to report the VM disk stats from the hosts. I have it for historical reasons, most probably not needed anymore if I manage to get the monitoring from the front-end working. Borrowing from your algorithm the idea to start from the images...

> Regarding monitoring, I have only issue with slow XPath when getting disks info from VM XML.

The xpath.rb is slow. In the latest refactoring (still work in progress) I am experimenting with `xmllint -xpath` which looks better, a python alternative for xpath.rb looks like a faster alternative too...
I am getting all VMs info in a temp file(`onevm list -x >temp_file`) and then extract needed VM's XML with xmllint
Will try to push something in a branch today or next week

**Comment by feldsam** [comment]
Hi, thanks for the info. I was playing with XPath a day ago. I figured out, that faster is running it on one VM XML that on all VMs XML. Probably in pythons is not much difference. Looking forward to seeing your python XmLlint example!

**Comment by atodorov-storpool** [comment]
The ugly python script is there from a long time https://github.com/OpenNebula/addon-storpool/blob/master/datastore/xpath_multi.py
I'am using it to extract the disks attributes with ';' as separator instead xpath.rb with %m% (space )
because if you have a volatile disk and after it an image, the CLONE attribute of the image is "shifted" to the volatile disk that has no CLONE attribute by definition

for example if you have

```xml
<DISK>
      <CLONE>NO</CLONE>
</DISK>
<DISK>
    <CLONE></CLONE>
</DISK>
<DISK>
   <CLONE>YES</CLONE>
</DISK>
```
with `xpath.rb %m%//DISK/CLONE` you will have 'NO  YES'  and when converting to array in bash you will have 
```bash
a=(NO  YES)
echo ${a[0]} ${a[1]} ${a[3]} 
'NO YES'
```
The last disk will be broken then....

**Comment by feldsam** [comment]
but it is a bug in OpenNebula core. Should be fixed to return CLONE = NO and not empty

**Comment by atodorov-storpool** [comment]
Raised the debug logging and updated to bb44f2b now I have in oned.log:

```
Fri May 22 18:06:35 2020 [Z0][ImG][D]: Message received: MONITOR SUCCESS 102 VVNFRF9NQj0yODA0OApUT1RBTF9NQj00Njk3MjIKRlJFRV9NQj00NDE2NzMKVk09W0lEPSI3IiwgUE9MTD0iIERJU0tfU0laRT1bSUQ9MCxTSVpFPTExMDNdIFNOQVBTSE9UX1NJWkU9W0lEPTAsRElTS19JRD0wLFNJWkU9Ml0KIl0K

Fri May 22 18:06:35 2020 [Z0][ImM][D]: Datastore SP-SYSTEM2 (102) successfully monitored.
Fri May 22 18:06:37 2020 [Z0][ImG][D]: Message received: MONITOR SUCCESS 100 VVNFRF9NQj0yODA0OApUT1RBTF9NQj00Njk3MjIKRlJFRV9NQj00NDE2NzMKVk09W0lEPSI2IiwgUE9MTD0iIERJU0tfU0laRT1bSUQ9MCxTSVpFPTExMDNdIERJU0tfU0laRT1bSUQ9MixTSVpFPTBdIERJU0tfU0laRT1bSUQ9MyxTSVpFPTBdIERJU0tfU0laRT1bSUQ9NCxTSVpFPTBdCiJdClZNPVtJRD0iNSIsIFBPTEw9IiBESVNLX1NJWkU9W0lEPTAsU0laRT0xOTY5XSBTTkFQU0hPVF9TSVpFPVtJRD0wLERJU0tfSUQ9MCxTSVpFPTddIFNOQVBTSE9UX1NJWkU9W0lEPTIsRElTS19JRD0wLFNJWkU9MTFdIERJU0tfU0laRT1bSUQ9MixTSVpFPTBdIERJU0tfU0laRT1bSUQ9MyxTSVpFPTBdIERJU0tfU0laRT1bSUQ9NCxTSVpFPTBdIERJU0tfU0laRT1bSUQ9NSxTSVpFPTBdCiJdCg==

Fri May 22 18:06:37 2020 [Z0][ImM][D]: Datastore SP-SYSTEM (100) successfully monitored.
```
The datastore info is updated but the VM's data not

**Comment by feldsam** [comment]
Do You know about this? https://github.com/OpenNebula/one/issues/2765

**Comment by atodorov-storpool** [comment]
Thanks for recalling this. It looks like it is not the same issue as there is no update at all.

```
grep MONITOR_VM_DISK /etc/one/oned.conf -A2
#  DS_MONITOR_VM_DISK: Number of MONIROTING_INTERVAL_DATASTORE intervals to monitor
#  VM disks. 0 to disable. Only applies to fs and fs_lvm datastores
#
--
#DS_MONITOR_VM_DISK = 10
```
The description is not very clear, tried with `DS_MONITOR_VM_DISK = 1` - no change...

Learned that in 5.12 the monitoring is split into two services. The shared datastores monitoring is in via DATASTORE_MAD (`-s` argument)
The ssh (non-shared) datastores are via a separate monitord service(using tm/monitor_ds on the hosts).

Can't make any of them working, though.

**Comment by feldsam** [comment]
It can be related to this commit https://github.com/OpenNebula/one/commit/44e44b47a859f03bac3177c879c7d1ff3f83c2b3

**Comment by feldsam** [comment]
Did you check/log if `/DS_DRIVER_ACTION_DATA/MONITOR_VM_DISKS` is passed?

**Comment by feldsam** [comment]
https://github.com/OpenNebula/one/blob/7dbed19430a041660b4e76c409940e1f8ac4880e/src/image/ImageManager.cc#L216

**Comment by paczerny** [comment]
I found bug in the disk monitoring on the host (local), I'm working on the fix.
After that I will fix the front-end monitoring

**Comment by atodorov-storpool** [comment]
looks like it is working. Need to figure out is there an issue in my monitoring regarding the new message format, though.

Here are some notes:
1. need dependency on _nc_ in the packaging. On CentOS7.8  _yum install nc_ installs _nmap-ncat_ 
2. the _nmap-ncat_ do not allow wait time '0'
3. In my setup, I've changed MONITOR_ADDRESS in /etc/monitord.conf so the monitor scripts should have an option to change this. 

The proposed patch expect environment variables to override the defaults

**Comment by feldsam** [comment]
Hello @atodorov-storpool , I don't see new format in monitor script. There is still `POOL` attr. Would you point me to code where you see new format?

**Comment by paczerny** [comment]
@feldsam Documentation for the new monitoring: http://docs.opennebula.io/5.11/deployment/open_cloud_host_setup/monitoring.html#extending-the-monitor-system

@atodorov-storpool We will address your issues with nc and MONITOR_ADDRESS soon

**Comment by feldsam** [comment]
@paczerny does this mean that we need rework tm_mad monitoring script? It looks like it is just for easier extending...

**Comment by atodorov-storpool** [comment]
@feldsam 
take a look at https://github.com/OpenNebula/one/blob/master/src/tm_mad/shared/monitor
the script sends the VM data directly to the new monitord. In other words, it is sending to two endpoints:
the datastore data to STDOUT (the "old" interface) and the VM disks data to the new monitord via netcat

**Comment by atodorov-storpool** [comment]
here is my ugly as hell refactoring (WIP) of the monitor script 
(There will be a configuration variable to handle both the classing and the new monitoring protocol)
https://github.com/OpenNebula/addon-storpool/blob/one-5.11.80-2/tm/storpool/monitor#L251

**Comment by atodorov-storpool** [comment]
The debugging of the new message format is driving my bash skills to a new level :)
```bash
printf "\x1f\x8b\x08\x00\x00\x00\x00\x00$(echo eJxz8Qz2jg/2jHK1jfZ0sTXQATMNDQ2MYxWC/RwDgj38Q5ClXUDKEQqNY7kAVwsRyAo=|base64 -d)" | gzip -dc -
```
(Yes, could be done in Ruby...)

**Comment by atodorov-storpool** [comment]
Confirming that the issue is fixed!

So we should expect _augeas_ as a dependency, then :)

**Comment by paczerny** [comment]
> Confirming that the issue is fixed!
Not yet, we have to rewrite it once more, as this solution doesn't cover all cases

**Comment by atodorov-storpool** [comment]
Ok! My master branch builder is monitoring GitHub twice per hour so I'll test it relatively soon after it is merged.

(btw the acceleration of the compiling is noticeable!)

**Comment by atodorov-storpool** [comment]
A feature request for testing?
Is it possible to log the decoded base64 of the MONITOR?
I am getting tired copy-paste to feed `base64 -d` ...

**Comment by feldsam** [comment]
Hello @atodorov-storpool , I revisited my tm/monitor script today. Replaced ruby xpath by xmlstarlet.

You can check https://github.com/OpenNebula/addon-3par/blob/master/tm/3par/monitor

**Comment by atodorov-storpool** [comment]
@feldsam 
Great! I believe there is no need to say that it is much faster :)

I've managed to make working monitor. Will update my work branch later, though.

**Comment by feldsam** [comment]
yes, it is, I have to solve multiple api calls to 3par, to even more speed up things.

**Comment by rsmontero** [comment]
Fixes in place. Please use this issue if additional changes are needed.

THANKS FOR YOUR FEEDBACK! :heart:

**Comment by atodorov-storpool** [comment]
> just fo reference, issue with vm disk monitoring was there until now on EL7 platforms. Fixed in this #5855

Still using our own implementation so missed/not affected/ by #5855
